File: atomic-x/android/src/main/kotlin/io/trtc/tuikit/atomicx/permission/Permission.kt
package io.trtc.tuikit.atomicx.permission

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Manages MethodChannel communication and delegates to PermissionHandler.
 */
class Permission(
    private val pluginBinding: FlutterPlugin.FlutterPluginBinding,
) : ActivityAware {

    companion object {
        private const val CHANNEL_NAME = "atomic_x/permission"
    }

    private val methodChannel: MethodChannel = MethodChannel(
        pluginBinding.binaryMessenger,
        CHANNEL_NAME,
    )
    private val handler: PermissionHandler = PermissionHandler(pluginBinding)
    private var activityBinding: ActivityPluginBinding? = null

    init {
        methodChannel.setMethodCallHandler(::onMethodCall)
    }

    private fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "requestPermissions" -> {
                call.argument<List<String>>("permissions")?.takeIf { it.isNotEmpty() }?.let {
                    handler.requestPermissions(it, result)
                } ?: result.error("INVALID_ARGUMENT", "Permissions are required", null)
            }
            "openAppSettings" -> result.success(handler.openAppSettings())
            "getPermissionStatus" -> {
                call.argument<String>("permission")?.let {
                    result.success(handler.getPermissionStatus(it))
                } ?: result.error("INVALID_ARGUMENT", "Permission is required", null)
            }
            else -> result.notImplemented()
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        attachActivity(binding)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        detachActivity()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        attachActivity(binding)
    }

    override fun onDetachedFromActivity() {
        detachActivity()
    }

    private fun attachActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        handler.setActivity(binding.activity)
        binding.addRequestPermissionsResultListener(handler)
    }

    private fun detachActivity() {
        activityBinding?.removeRequestPermissionsResultListener(handler)
        handler.setActivity(null)
        activityBinding = null
    }

    fun dispose() {
        detachActivity()
        methodChannel.setMethodCallHandler(null)
        handler.dispose()
    }
}

File: atomic-x/ios/Classes/Permission/Permission.swift
import Flutter

/// Manages MethodChannel communication and delegates to PermissionHandler.
class Permission {
  private static let channelName = "atomic_x/permission"
  private let methodChannel: FlutterMethodChannel
  private let registrar: FlutterPluginRegistrar
  private let handler: PermissionHandler

  init(registrar: FlutterPluginRegistrar) {
    self.registrar = registrar
    self.handler = PermissionHandler()
    self.methodChannel = FlutterMethodChannel(
      name: Permission.channelName,
      binaryMessenger: registrar.messenger()
    )
    
    // Set method call handler
    methodChannel.setMethodCallHandler { [weak self] (call, result) in
      self?.handleMethodCall(call, result: result)
    }
  }

  private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "requestPermissions":
      guard let args = call.arguments as? [String: Any],
            let permissions = args["permissions"] as? [String] else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Permissions are required", details: nil))
        return
      }
      handler.requestPermissions(permissions, result: result)

    case "openAppSettings":
      result(handler.openAppSettings())

    case "getPermissionStatus":
      guard let args = call.arguments as? [String: Any],
            let permission = args["permission"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENT", message: "Permission is required", details: nil))
        return
      }
      result(handler.getPermissionStatus(permission))

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
